### DISCUSSION TITLE
Why Symfony does not have HTTP middlewares?

### DISCUSSION THREAD

**Discussion by gnumoksha** [symfony/symfony#49531] [Q&A] [answered]
Why [PSR-15 middlewares](https://www.php-fig.org/psr/psr-15/), or any other kind of HTTP middlewares, were not implemented?

**Comment by GromNaN** [comment, marked as answer]
There is [stackphp](https://stackphp.com/), created 10 years ago. It never reached a large adoption, I think it is because event listeners are a lot more powerful: you can hook in a [lot of kernel events](https://symfony.com/doc/current/components/http_kernel.html) and read/write various data into the request/response.

If you want to use the PSR-15 middlewares, it may be possible to use the [Symfony PSR-7 bridge](https://symfony.com/doc/current/components/psr7.html).
